[Beck] step1 - Start IPad App Project using MVC Pattern

SangHwi-Back opened the pull request:
> 기존의 PR 메시지는 루카스의 내용이 너무 많이 들어가서 삭제하게 되었습니다.

## 작업 목록

* 크기가 고정되고 위치는 랜덤인 사각형 뷰를 위한 Model 객체를 생성하였습니다. 모델 역할만을 하기 때문에, 뷰와 관련된 Core Graphics, UIKit 등의 프레임워크는 임포트 하지 않습니다.
* 주어진 조건에 따라 랜덤 값을 지정하여 반환해주는 팩토리 객체와 메소드를 생성하였습니다.
* 팩토리 메소드가 사용할 랜덤 값 추출 관련 객체를 따로 분리하여 역할을 나누었습니다. 재사용에 용이할 것이라고 판단하였습니다.
* 객체를 구현할 때 테스트에 용이하게 만들기 위해서 print 사용을 자제하고 CustomStringConvertible 프로토콜을 채택하였습니다.
* 시스템 로그를 실습해보았습니다.

## 학습 키워드

```MVC Pattern``` ```Model``` ```OOP``` ```XCTest``` ```Factory Pattern```

## PR 재요청을 위해 가장 고심한 부분

팩토리 클래스와 메소드가 각각의 역할이 있는데 이 구분이 명확하지 않았던 점이 가장 문제였습니다. 팩토리 클래스를 사용하고는 있지만, 내부의 메소드는 명확한 역할이 구분되지 않아서 소스를 나누는데 집중하였습니다.

모델에 뷰와 관련된 부분을 삽입한 것은 명백히 잘못된 부분이었습니다. 모두 제외하였습니다.

모델의 상위 객체인 뷰 컨트롤러와 통신하기 위해 슈퍼 뷰를 참조하는 점을 수정하였습니다. 델리게이트 패턴을 사용해서 모델 클래스의 유연성을 높이고 직접적인 참조관계가 성립하지 않도록 프로토콜을 추가하고 구현하였습니다.

SangHwi-Back committed:
DrawingApp Initial Commit

드로잉 앱 생성 후 최초 커밋합니다.

SangHwi-Back committed:
Configure App Setting and README ref

README 파일 레퍼런스를 추가하고
애플리케이션을 가로 모드만 사용하는 아이패드 앱으로 설정한다.

SangHwi-Back committed:
Program work at README

미리 README에 해야 할 작업들을 정의합니다.
MVC와 객체지향을 연결하여 앱을 생성하고 뷰를 생성하는 모델을 지금까지 정의한 방식과는
확연히 다르게 생성하는 방식을 실습해 볼 것입니다.

SangHwi-Back committed:
Add UI for randomView

random하게 뷰의 프로퍼티 모델을 생성하는 팩토리 메소드를 생성합니다.
팩토리 메소드는 랜덤한 값들을 생성하는 내부 메소드를 가지고 있으며,
뷰 컨트롤러는 팩토리 메소드를 통해 랜덤 뷰 프로퍼티들을 쉽게 얻어올 수 있습니다.
앞으로 접근제한자 처리, 역할 분담, 테스트에 맞게 메소드 수정할 예정입니다.

SangHwi-Back committed:
SubClass to Factory method

랜덤 값을 생성하는 클래스, 최대값을 받아 랜덤 타입의 튜플을 생성하는 클래스,
그리고 튜플들을 특정 타입으로 저장하여 뷰의 모델 데이터로 사용할 수 있는 클래스로 모델을 분류하였습니다.

SangHwi-Back committed:
Refactor Model Class

모델 클래스를 업데이트 하여,  테스트에 적합하도록 변경하였습니다.
뷰의 크기와 위치를 무작위로 추출하는 함수와 뷰의 위치를 무작위로 추출(크기는 고정)하는 함수를 나눠
각각 테스트할 수 있도록 하였습니다.
뷰가 화면을 벗어나는지 테스트하기 위해 FactoryViewRandomProperty.isExceededSuperView() 를 만들었습니다.

SangHwi-Back committed:
Add SystemLog Utility

시스템 로그를 남기는 것에 대해 학습하던 중 유용한 유틸리티용 코드를 찾았습니다.
이를 그대로 적용하여 실습해 보았고, 내용을 간단하게나마 정리하여 README 추가한 후 PR 전송하겠습니다.

SangHwi-Back committed:
Update README about Step1

스텝1을 수행하면서 작업한 내용과 추가 학습 내용으로 주어진 내용을 정리하였습니다.

SangHwi-Back committed:
Refactor Before Pull Request

풀 리퀘스트 이전에 접근제한자, 객체 역할 변경, 테스트 수정 등을 진행하였습니다.

godrm commented:
PR 작업 목록을 작성할 때 루카스 요구사항을 그대로 복사하지 마시고,
스스로 작업 단위를 나누고, 요약해서 기록하면 됩니다.

godrm commented on DrawingApp/DrawingApp/FactoryViewRandomProperty.swift:
```diff
@@ -0,0 +1,62 @@
+//
+
+import Foundation
+import UIKit
+
+class FactoryViewRandomProperty: ViewPropertyCreator {
+    
+    private let name: String
+    private let superview: UIView
+    
+    private var viewFrame: CGRect {
+        superview.frame
+    }
```
이번 단계가 UIKit 프레임워크를 임포트해서 사용하지 않고 구현하는 건데 UIKit을 임포트하고 있네요
UIView도 있고 뷰 객체를 벌써 만들고 계신건가요? 뷰 없이 Model 부터 선언하고 확인하는 단계였던 것 같은데요?

SangHwi-Back replied:
정신없이 진행하다 보니 요구사항을 벗어난 것 같습니다. 다시 Push 하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/FactoryViewRandomProperty.swift:
```diff
@@ -0,0 +1,62 @@
+    init(name: String, superview: UIView) {
+        
+        self.name = name
+        self.superview = superview
```
이 타입이 뷰인지 모델인지 모르겠지만, 그렇더라도 superview를 참조할 필요가 없습니다
항상 포함관계에서 하위에 있는 타입은 자신을 포함하는 상위 타입을 모르는 게 더 좋습니다

godrm commented on DrawingApp/DrawingApp/FactoryViewRandomProperty.swift:
```diff
@@ -0,0 +1,62 @@
+            size: locationProperty.size,
+            color: rectRGBPointGenerator(maxR: 255, maxG: 255, maxB: 255),
+            alpha: Double.random(in: 1.0...10.0)
+        )
```
특정한 값으로 인스턴스를 생성하는 메소드를 모아서 팩토리로 분리하는 거였는데, 여기서는 자기가 자신의 인스턴스를 생성하고 있네요
그러면 init()과 어떤 차이가 있는걸까요?

SangHwi-Back replied:
@godrm 🙋 안녕하세요 JK! 말씀해주신 코멘트를 작업 중 질문드립니다.

제가 최초 작성한 코드는 Model객체의 init()을 호출하는 부분이 뷰 컨트롤러에서 init을 호출하는 것과 다르지 않아서 팩토리 메소드 같지는 않다고 말씀해주시는 것으로 이해했습니다.

개선할 방법을 아래와 같이 생각하고 있는데, 아직도 제가 잘못 이해하고 있는건지 헷갈려서 다시 한번 질문드립니다. 혹시나 틀린 부분이 있다면 지적을 부탁드립니다. 감사합니다! 

1. 무작위로 값을 뽑는 부분을 팩토리 클래스의 Function 혹은 Property(모두 private)로 선언하고 팩토리 메소드에서 사용한다.
2. 팩토리 클래스의 서브 타입으로, 무작위 값 추출에 필요한 최대값만 저장하는 값 타입을 선언합니다. Delegate 패턴을 이용하여 상위 객체에서 받아옵니다.
3. 팩토리 메소드는 클래스에 미리 지정된 Fucntion, Property, Delegate 객체만을 이용해서 Model 객체를 만들어 냅니다.

godrm replied:
복잡하게 생각하지마시구요
- 어떤 타입 A가 있고, A.init()으로 초기화할 수 있어야 합니다
- A에 대한 팩토리는 A.init()을 대신 호출해주면서 복잡한 값들을 간단하게 넣고 생성할 수 있으면 됩니다
팩토리에서 뭘하더라도 A.init() 역할이 바뀌는 건 없습니다

SangHwi-Back replied:
넵 알겠습니다. 다시 시도해보겠습니다.... 감사합니다!

godrm commented on DrawingApp/DrawingApp/String+Extension.swift:
```diff
@@ -0,0 +1,34 @@
+    static var allAlphabet = (uppercaseAlphabet + lowercaseAlphabet)
+
+    static var randomViewIdGenerator: String {
+//        let char = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+//        var result = String((0..<3).map{ _ in char.randomElement()!})
+//        for _ in 0..<2 {
+//            result += ("-"+String((0..<3).map{ _ in char.randomElement()!}))
+//        }
+//        return result
+        
```
사용하지 않는 코드는 주석으로 남기지말고 삭제하시는 게 좋습니다

godrm commented on DrawingApp/DrawingApp/String+Extension.swift:
```diff
@@ -0,0 +1,34 @@
+        for _ in 0..<2 {
+            result += "-\(randomThreeAlphabet)"
+        }
+        return result
```
이런 Id를 생성하는 메소드를 String에 확장할 때는 모든 String이 다 사용하는 메소드인지 판단해보고 확장하는 게 좋습니다
특정 타입이나 특정 함수에서만 쓴다면 그것과 관련된 곳에 구현하는 게 더 좋습니다

godrm commented on DrawingApp/DrawingApp/SystemLog.swift:
```diff
@@ -0,0 +1,120 @@
+
+    static func custom(category: String, _ message: Any, _ arguments: Any...) {
+        log(message, arguments, level: .custom(categoryName: category))
+    }
```
로그 라이브러이를 만드셨네요. Logger라는 타입도 확인해보세요

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -0,0 +1,40 @@
+
+    @IBAction func addRectButtonTouchUpInside(_ sender: UIButton) {
+        let subviewName = "subView #\(view.subviews.count)"
+        let viewModel = FactoryViewRandomProperty(name: subviewName, superview: view).make()
+        
+        let randomView = UIView(frame: CGRect.useViewModel(point: viewModel.point, size: viewModel.size))
+        randomView.backgroundColor = UIColor.useViewModel(rgbValue: viewModel.rgbValue, alpha: viewModel.alpha)
+        
+        view.addSubview(randomView)
```
이번 단계 요구사항을 오해하신게 아닌가 싶습니다
뷰를 만드는 게 아니라, 뷰를 처리하기 위해 값을 가지는 모델을 선언하고 값을 확인하는 겁니다
이렇게 뷰만 만드는 것을 하지 않기 위해서 나눈 단계입니다. 다시 확인해봐주세요

SangHwi-Back replied:
아... 다시한번 자세히 읽어보고 실수를 깨달았습니다. 수정 후 재요청 드리도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewPropertyCreator.swift:
```diff
@@ -0,0 +1,35 @@
+// 랜덤 값을 추출하는 RandomizeValue 를 서브클래싱 합니다.
+class ViewPropertyCreator: RandomizeValue {
+    
+    func rectPointGenerator(maxPointX: Double, maxPointY: Double) -> RectPoint {
+        (
```
함수 이름이 동사가 아닌 것 같습니다. 이걸 읽으면 Generator를 위한 좌표를 네모로 만드는 듯한 느낌입니다.

godrm commented on DrawingApp/DrawingApp/ViewRandomProperty.swift:
```diff
@@ -0,0 +1,49 @@
+import UIKit
+
+// typeAlias를 통해 튜플의 용도가 이름에서도 드러나도록 하였습니다.
+typealias RectSize = (width:Double, height:Double)
+typealias RectPoint = (x:Double, y:Double)
+typealias RectRGBColor = (r:Double, g:Double, b: Double)
```
이제는 튜플+alias 보다는 구체적인 값 타입을 지정해보세요

godrm commented on DrawingApp/DrawingApp/ViewRandomProperty.swift:
```diff
@@ -0,0 +1,49 @@
+    let point: RectPoint
+    
+    let rgbValue: RectRGBColor
+    let alpha: Double
```
alpha에 대한 범위 요구사항이 빠진 것 같습니다

SangHwi-Back replied:
1-10 *단계* 였군요.... 이번 코드는 실수가 많아 죄송합니다. 다시 Push 후 PR 하도록 하겠습니다.

godrm reviewed (changes requested):
이번 단계 요구사항에서 조금 벗어난 부분이 보입니다.
피드백과 요구사항을 다시 한 번 확인해보시고 다시 보내봐주세요. 
뷰를 만드는 게 아니라 모델을 만들어야 하는데 그 부분을 확인하기 어려웠습니다

SangHwi-Back committed:
Correct Step1 mistakes following Comment

코멘트에서 지적받은 사항에 대해 아래와 같이 수정하였습니다.
1. 뷰를 만드는 것은 현재 스텝이 아니므로 제외합니다.
2. 모델은 UIKit 사용이 금지되어 있습니다. MVC 패턴의 모델은 뷰와 상관없습니다.
3. 함수 이름은 동사를 사용하는 것이 좋습니다.

SangHwi-Back committed:
Correct Step1 complement following Comment

코멘트에서 지적받은 사항에 대해 아래와 같이 수정하였습니다.
1. Tuple이었던 typealias를 struct로 바꾸어 보았습니다.
2. 불필요한 코드, 이번 스텝에 벗어난 코드를 전부 삭제하였습니다.
3. Naming이 애매한 부분을 찾아 수정하였습니다.

SangHwi-Back committed:
Correct Step1 Logger task

코멘트에서 추가 학습거리인 Logger 구조체를 실습해 보았습니다.

godrm commented on DrawingApp/DrawingApp/ViewRandomProperty.swift:
```diff
@@ -6,14 +6,27 @@
+    var r: Double
+    var g: Double
+    var b: Double
+}
```
타입이 다르면 파일로 분리해보세요. 여기 Size나 Point가 꼭 Rect 와 관련된 거라면 Nested 타입도 고려해보시구요

SangHwi-Back replied:
Nested 타입을 썼을 때와 쓰지 않았을 때를 고려해보고 Model 소스코드에서 분리하도록 하겠습니다. 감사합니다!
